Source/Core/Core/AchievementManager.cpp

```diff
@@ -1016,694 +520,461 @@ void AchievementManager::FilereaderClose(void* file_handle)
+  INFO_LOG_FMT(ACHIEVEMENTS, "Reset requested by Achievement Mananger");
+  Core::Stop(Core::System::GetInstance());
 }
 
-void AchievementManager::HandleAchievementUnprimedEvent(const rc_runtime_event_t* runtime_event)
+void AchievementManager::HandleServerErrorEvent(const rc_client_event_t* client_event)
 {
-  if (!Config::Get(Config::RA_BADGES_ENABLED))
-    return;
-  auto it = m_unlock_map.find(runtime_event->id);
-  if (it == m_unlock_map.end())
-  {
-    ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid achievement unprimed event with id {}.",
-                  runtime_event->id);
-    return;
-  }
-  m_active_challenges.erase(it->second.unlocked_badge.name);
+  ERROR_LOG_FMT(ACHIEVEMENTS, "RetroAchievements server error: {} {}",
+                client_event->server_error->api, client_event->server_error->error_message);
 }
 
-void AchievementManager::HandleLeaderboardStartedEvent(const rc_runtime_event_t* runtime_event)
+static std::unique_ptr<OSD::Icon> DecodeBadgeToOSDIcon(const AchievementManager::Badge& badge)
 {
-  for (u32 ix = 0; ix < m_game_data.num_leaderboards; ix++)
+  if (badge.empty())
+    return nullptr;
+
+  auto icon = std::make_unique<OSD::Icon>();
+  if (!Common::LoadPNG(badge, &icon->rgba_data, &icon->width, &icon->height))
   {
-    if (m_game_data.leaderboards[ix].id == runtime_event->id)
-    {
-      OSD::AddMessage(fmt::format("Attempting leaderboard: {}", m_game_data.leaderboards[ix].title),
-                      OSD::Duration::VERY_LONG, OSD::Color::GREEN);
-      return;
-    }
+    ERROR_LOG_FMT(ACHIEVEMENTS, "Error decoding badge.");
+    return nullptr;
   }
-  ERROR_LOG_FMT(ACHIEVEMENTS, "Invalid leaderboard started event with id {}.", runtime_event->id);
+  return icon;
 }
 
-void AchievementManager::HandleLeaderboardCanceledEvent(const rc_runtime_event_t* runtime_event)
+void AchievementManager::Request(const rc_api_request_t* request,
+                                 rc_client_server_callback_t callback, void* callback_data,
+                                 rc_client_t* client)
 {
-  for (u32 ix = 0; ix < m_game_data.num_leaderboards; ix++)
-  {
-    if (m_game_data.leaderboards[ix].id == runtime_event->id)
+  std::string url = request->url;
+  std::string post_data = request->post_data;
+  AchievementManager::GetInstance().m_queue.EmplaceItem([url, post_data, callback, callback_data] {
+    const Common::HttpRequest::Headers USER_AGENT_HEADER = {{"User-Agent", "Dolphin/Placeholder"}};
+
+    Common::HttpRequest http_request;
+    Common::HttpRequest::Response http_response;
+    if (!post_data.empty())
+      http_response = http_request.Post(url, post_data, USER_AGENT_HEADER,
+                                        Common::HttpRequest::AllowedReturnCodes::All);
+    else
+      http_response =
+          http_request.Get(url, USER_AGENT_HEADER, Common::HttpRequest::AllowedReturnCodes::All);
+
+    std::string error_message = "Failed HTTP request.";
```

Comment:
At the very least, please move this line into the `else` block below to avoid the dynamic allocation as a constant overhead for the function.

I'd also consider avoiding the allocation of the `error_message` in favor of something like:

```
constexpr char error_message[] = "Failed HTTP request.";
server_response.body = error_message;
server_response.body_length = sizeof(error_message);
```